core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -312,18 +314,13 @@ protected final void requestMessagesFromDeframer(int numMessages) {
    * StreamListener#onReady()} handler if appropriate. This must be called from the transport
    * thread, since the listener may be called back directly.
    */
-  protected void onStreamAllocated() {
-    boolean doNotify;
+  protected final void onStreamAllocated() {
+    checkState(listener() != null);
     synchronized (onReadyLock) {
-      if (allocated) {
-        throw new IllegalStateException("Already allocated");
-      }
+      checkState(!allocated, "Already allocated");
       allocated = true;
-      doNotify = needToNotifyOnReady();
```

Comment:
So far it seems like neither listener nor allocated is locked in the original version (though it seems like that would be a good thing to do).  Partially threadsafe classes frighten me.